include/envoy/registry/registry.h

```diff
@@ -18,6 +18,51 @@
 namespace Envoy {
 namespace Registry {
 
+template <class Base> class FactoryRegistry;
+
+/**
+ * Helper class to call `registeredNames` for a specialized
+ * FactoryRegistry.
+ */
+class BaseFactoryCategoryNames {
+public:
+  virtual ~BaseFactoryCategoryNames() = default;
+  virtual std::vector<absl::string_view> registeredNames() const PURE;
+};
+
+template <class Base> class FactoryCategoryNames : public BaseFactoryCategoryNames {
+public:
+  using FactoryRegistry = Envoy::Registry::FactoryRegistry<Base>;
+
+  std::vector<absl::string_view> registeredNames() const override {
+    return FactoryRegistry::registeredNames();
+  }
+};
+
+/**
+ * FactoryCategoryRegistry registers factory registries by their
+ * declared category. The category is exposed by a static category()
+ * method on the factory base type.
+ */
+class FactoryCategoryRegistry {
+public:
+  using MapType = absl::flat_hash_map<std::string, BaseFactoryCategoryNames*>;
+
+  static MapType& factories() {
```

Comment:
nit: can the non-const method be private and then there is a public one that the server.cc code calls?

Reply:
Fixed.